sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/OutboundObserverFactory.java

```diff
@@ -0,0 +1,138 @@
+ */
+public abstract class OutboundObserverFactory {
+  /**
+   * Create a buffering {@link OutboundObserverFactory} for client-side RPCs with the specified
+   * {@link ExecutorService} and the default buffer size.
+   */
+  public static OutboundObserverFactory clientBuffered(ExecutorService executorService) {
+    return new Buffered(executorService, Buffered.DEFAULT_BUFFER_SIZE);
+  }
+
+  /**
+   * Create a buffering {@link OutboundObserverFactory} for client-side RPCs with the specified
+   * {@link ExecutorService} and buffer size.
+   */
+  public static OutboundObserverFactory clientBuffered(
+      ExecutorService executorService, int bufferSize) {
+    return new Buffered(executorService, bufferSize);
+  }
+
+  /**
+   * Create the default {@link OutboundObserverFactory} for client-side RPCs, which uses basic
+   * unbuffered flow control and adds synchronization to provide thread safety of access to the
+   * returned observer.
+   */
+  public static OutboundObserverFactory clientDirect() {
+    return new DirectClient();
+  }
+
+  /** Like {@link #clientDirect} but for server-side RPCs. */
+  public static OutboundObserverFactory serverDirect() {
+    return new DirectServer();
+  }
+
+  /**
+   * Creates an {@link OutboundObserverFactory} that simply delegates to the base factory, with no
+   * flow control or synchronization. Not recommended for use except in tests.
+   */
+  public static OutboundObserverFactory trivial() {
+    return new Trivial();
+  }
+
+  /**
+   * Creates an outbound observer for the given inbound observer by potentially inserting hooks into
+   * the inbound and outbound observers.
+   *
+   * @param baseOutboundObserverFactory A base function to create an outbound observer from an
+   *     inbound observer.
+   * @param inboundObserver The inbound observer.
+   */
+  public abstract <ReqT, RespT> StreamObserver<RespT> outboundObserverFor(
+      Function<StreamObserver<ReqT>, StreamObserver<RespT>> baseOutboundObserverFactory,
+      StreamObserver<ReqT> inboundObserver);
+
+  private static class DirectClient extends OutboundObserverFactory {
+    @Override
+    public <ReqT, RespT> StreamObserver<RespT> outboundObserverFor(
+        Function<StreamObserver<ReqT>, StreamObserver<RespT>> baseOutboundObserverFactory,
+        StreamObserver<ReqT> inboundObserver) {
+      AdvancingPhaser phaser = new AdvancingPhaser(1);
+      inboundObserver = ForwardingClientResponseObserver.create(inboundObserver, phaser::arrive);
+      CallStreamObserver<RespT> outboundObserver =
+          (CallStreamObserver<RespT>) baseOutboundObserverFactory.apply(inboundObserver);
+      return new DirectStreamObserver<>(phaser, outboundObserver);
+    }
+  }
+
+  private static class DirectServer extends OutboundObserverFactory {
+    @Override
+    public <ReqT, RespT> StreamObserver<RespT> outboundObserverFor(
+        Function<StreamObserver<ReqT>, StreamObserver<RespT>> baseOutboundObserverFactory,
```

Comment:
Can we keep the `StreamObserverClientFactory` or an equivalently named type around so we can gave contextual meaning as to what we want and then the javadoc on the class can give further explanation instead of using a Function?